docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,436 @@
+  * [Design Details](#design-details)
+    * [Test Plan](#test-plan)
+    * [Graduation Criteria](#graduation-criteria)
+    * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    * [Version Skew Strategy](#version-skew-strategy)
+  * [Implementation History](#implementation-history)
+  * [Drawbacks](#drawbacks)
+  * [Alternatives](#alternatives)
+      * [Object References, Templates, MachineSets and MachineDeployments](#object-references-templates-machinesets-and-machinedeployments-1)
+
+## Glossary
+
+- **Cluster API**: Unless otherwise specified, this refers to the project as a whole.
+- **Machine**: the Custom Resource for Kubernetes
+- **Server/Instance/Host**: the infrastructure that backs a Machine
+- **Bootstrapping**: turning a server into a Kubernetes node
+
+## Summary
+
+This proposal outlines splitting up the Machine data model into two new providers/controllers along with the generic Machine controller. The two new providers are the Infrastructure provider and the Bootstrap provider. The infrastructure provider handles provisioning infrastructure (cloud, bare metal, virtual) and the bootstrap provider handles turning a server into a Kubernetes node. This change will improve separation of concerns for the Machine controller and gives more degrees of freedom to the user.
+
+In Cluster API v1alpha1, users can create Machine custom resources. When a Machine resource is created, Cluster API components react by provisioning infrastructure (typically a physical or virtual machine) and then bootstrapping Kubernetes on the provisioned server, making it a Kubernetes Node. Cluster API’s current architecture mandates that a single “provider” handle both the infrastructure provisioning and the Kubernetes bootstrapping.
+
+## Motivation
+- A singular provider that combines infrastructure provisioning and Kubernetes bootstrapping is too restrictive. Many current v1alpha1 providers target single cloud providers. If a user wants to install a specific Kubernetes distribution (Kubernetes, OpenShift, Rancher, etc.), it is left to the provider to add support for it. In practical terms, we are not able to guarantee that a user can install the same distribution regardless of infrastructure environment.
+- Both users and developers are confused by the code organization split where all the controllers live in the cluster-api repository, but some controllers are part of the generic cluster-api pod (MachineDeployment, MachineSet), while other controllers are part of each provider’s pod (Cluster, Machine). We would like for the code location to correspond to the pod, which hopefully will reduce both user and developer confusion.
+- The Machine custom resource contains fields for provider-specific configuration and status. These are represented as opaque [RawExtensions](https://godoc.org/k8s.io/apimachinery/pkg/runtime#RawExtension), meaning their contents are not directly processed as part of the Machine’s validation.
+
```

Comment:
The new separation might make support for management of multiple cloud providers in the same cluster easier/trivial. Right now, machine controller is bound to a specific cloud provider. All machines watched by the controller are considered homogeneous. With the new layout one can deploy just one machine controller. Presence of a specific infrastructure provider will decide if a machine gets into a running state eventually or not.

@vincepri can we extend the list of motivation items with another one?:
```
Code integration of machine controller with provider specific actuator does not allow to simply reconcile machines that have provider spec of different cloud providers in the same cluster (i.e. hybrid cluster). Separation of provider specific bits from the machine controller will increase flexibility of the cluster api plane and move responsibility for their interpretation and reconciliation into dedicated controller(s).
```

Reply:
Added, thank you!